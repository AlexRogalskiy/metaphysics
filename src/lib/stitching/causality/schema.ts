import { createCausalityLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
  FilterRootFields,
  FilterTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

const blacklistedTypes: string[] = []
const whitelistedRootFields: string[] = []
const privateFields: string[] = ["lotStandings"]
const permittedRootFields = [...whitelistedRootFields, ...privateFields]

export const executableCausalitySchema = () => {
  const causalityLink = createCausalityLink()
  const causalityTypeDefs = readFileSync("src/data/causality.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: causalityTypeDefs,
    link: causalityLink,
    buildSchemaOptions: {
      assumeValidSDL: true,
    },
  })

  // Return the new modified schema
  return transformSchema(schema, [
    new FilterTypes((type) => !blacklistedTypes.includes(type.name)),
    new FilterRootFields((_operation, name, _field) => {
      return permittedRootFields.includes(name)
    }),
    new RenameTypes((name) => {
      // this could be left as AuctionsLong, it is just a scalar (same as int?)
      // we should verify that JSON + javascript can hold all possible values
      // we would send (at least 800 billion * 100)
      // also we need to manually add this to the schema - figure out why it isn't
      // exported
      if (name === "Long") return "Long"
      if (name === "Lot") name = "LotState"
      if (name === "Sale") name = "SaleState"
      return `Auctions${name}`
    }),
    new RenameRootFields((_operation, name) => {
      const capitalizedName = name.charAt(0).toUpperCase() + name.slice(1)
      if (privateFields.includes(name)) {
        return `_unused_auctions${capitalizedName}`
      }
      return `auctions${capitalizedName}`
    }),
  ])
}
