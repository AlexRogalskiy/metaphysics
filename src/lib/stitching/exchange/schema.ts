import { createStressLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableStressSchema = async () => {
  const stressSDL = readFileSync("src/data/exchange.graphql", "utf8")
  const stressLink = createStressLink()

  const schema = await makeRemoteExecutableSchema({
    schema: stressSDL,
    link: stressLink,
  })

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      return `Ecommerce${name}`
    }),
    new RenameRootFields((_operation, name) => `ecommerce_${name}`),
  ])
}
