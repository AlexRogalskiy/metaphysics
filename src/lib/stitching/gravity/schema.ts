import { createGravityLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  FilterTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableGravitySchema = () => {
  const gravityTypeDefs = readFileSync("src/data/gravity.graphql", "utf8")

  const gravityLink = createGravityLink()
  const schema = makeRemoteExecutableSchema({
    schema: gravityTypeDefs,
    link: gravityLink,
  })

  // Types which come from Gravity which MP already has copies of.
  // In the future, these could get merged into the MP types.
  const removeTypes = [
    "Artist",
    "Artwork",
    "ArtistEdge",
    "ArtworkEdge",
    "ArtworkConnection",
    "ArtistConnection",
    "Partner",
  ]

  // Return the new modified schema
  return transformSchema(schema, [
    // Remove types which Metaphysics handles better
    new FilterTypes(type => {
      return !removeTypes.includes(type.name)
    }),
  ])
}
