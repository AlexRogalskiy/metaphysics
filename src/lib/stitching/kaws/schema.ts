import { createKawsLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableKawsSchema = () => {
  const kawsLink = createKawsLink()
  const kawsTypeDefs = readFileSync("src/data/kaws.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: kawsTypeDefs,
    link: kawsLink,
  })

  // Remap the names of certain types from kaws to fit in the larger
  // metaphysics ecosystem.
  const remap = {
    Collection: "MarketingCollection",
    CollectionQuery: "MarketingCollectionQuery",
    Image: "MarketingImage",
  }

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      const newName = remap[name] || name
      return newName
    }),
    new RenameRootFields(
      (_operation, name) =>
        `marketing${name.charAt(0).toUpperCase() + name.slice(1)}`
    ),
  ])
}
