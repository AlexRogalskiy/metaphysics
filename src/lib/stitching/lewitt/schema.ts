import { createLewittLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableLewittSchema = async () => {
  const lewittSDL = readFileSync("src/data/lewitt.graphql", "utf8")
  const lewittLink = createLewittLink()

  const schema = await makeRemoteExecutableSchema({
    schema: lewittSDL,
    link: lewittLink,
  })

  // Remap the names of certain types from Lewitt to fit in the larger
  // metaphysics ecosystem.
  const remap = {
    Currencies: "PartnerProductInvoiceCurrencies",
    Invoice: "PartnerProductInvoice",
    CreateInvoiceInput: "PartnerProductCreateInvoiceInput",
    ArtworkGroup: "PartnerProductInvoiceArtworkGroup",
    LineItem: "PartnerProductInvoiceLineItem",
    MerchantAccount: "PartnerProductInvoiceMerchantAccount",
    Json: "PartnerProductJson",
  }

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      const newName = remap[name]
      if (!newName) {
        throw new Error(
          `All types inside Lewitt should be mapped.\n Missing ${name}`
        )
      }
      return newName
    }),
  ])
}
