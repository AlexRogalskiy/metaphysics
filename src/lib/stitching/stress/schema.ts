import { createStressLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableStressSchema = async () => {
  const stressSDL = readFileSync("src/data/stress.graphql", "utf8")
  const stressLink = createStressLink()

  const schema = await makeRemoteExecutableSchema({
    schema: stressSDL,
    link: stressLink,
  })

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      return `Ecommerce${name}`
    }),
    // new TransformRootFields((operation, name, field) => {
    //   console.log(operation, name)
    //   if (operation == "Query") {
    //     return { name: `ecommerce_${name}`, field }
    //   } else {
    //     return { name, field }
    //   }
    // }),
  ])
}
