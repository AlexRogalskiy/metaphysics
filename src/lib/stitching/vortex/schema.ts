import { createVortexLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableVortexSchema = () => {
  const vortexLink = createVortexLink()
  const vortexTypeDefs = readFileSync("src/data/vortex.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: vortexTypeDefs,
    link: vortexLink,
  })

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      return `Analytics${name}`
    }),
    new RenameRootFields(
      (_operation, name) =>
        `analytics${name.charAt(0).toUpperCase() + name.slice(1)}`
    ),
  ])
}
