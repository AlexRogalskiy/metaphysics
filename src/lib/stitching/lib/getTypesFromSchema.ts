import { graphql, GraphQLSchema } from "graphql"
import gql from "lib/gql"

export const getTypesFromSchema = async (schema: GraphQLSchema) => {
  // An introspection query that pulls out the names of all types in the schema
  const q = gql`
    {
      __schema {
        types {
          name
        }
      }
    }
  `
  const response = await graphql(schema, q, null, null)
  // Throw an error to make it easy to read in the failing test
  if (response.errors) {
    throw new Error(
      `Got errors from GQL: ${JSON.stringify(response.errors, null, "  ")}`
    )
  }

  // Return the real data
  return response.data!.__schema.types.map(t => t.name)
}
