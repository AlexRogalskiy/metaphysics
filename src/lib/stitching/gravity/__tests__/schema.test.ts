import { executableGravitySchema } from "../schema"
import { getTypesFromSchema } from "lib/stitching/lib/getTypesFromSchema"

it("Does not include blacklisted types", async () => {
  const gravitySchema = await executableGravitySchema()
  const gravityTypes = await getTypesFromSchema(gravitySchema)

  expect(gravityTypes).not.toContain("Artist")
  expect(gravityTypes).not.toContain("Artwork")
  expect(gravityTypes).not.toContain("Partner")
})
