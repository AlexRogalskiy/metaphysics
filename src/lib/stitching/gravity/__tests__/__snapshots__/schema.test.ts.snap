// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates an SDL 1`] = `
"type DoNotUseThisPartner {
  displayName: String
  display_name: String @deprecated(reason: \\"Use displayName\\")
  givenName: String
  given_name: String @deprecated(reason: \\"Use givenName\\")
  id: ID!
  relativeSize: Int
  relative_size: Int @deprecated(reason: \\"Use relativeSize\\")
  slug: String
  subscriptionState: String
  subscription_state: String @deprecated(reason: \\"Use subscriptionState\\")
}

type Mutation {
  recordArtworkView(input: RecordArtworkViewInput!): RecordArtworkViewPayload
}

# Information about pagination in a connection.
type PageInfo {
  # When paginating forwards, the cursor to continue.
  endCursor: String

  # When paginating forwards, are there more items?
  hasNextPage: Boolean!

  # When paginating backwards, are there more items?
  hasPreviousPage: Boolean!

  # When paginating backwards, the cursor to continue.
  startCursor: String
}

type Query {
  # Autocomplete resolvers.
  _unused_gravity_matchPartners(matchType: String, page: Int = 1, size: Int = 5, term: String!): [DoNotUseThisPartner!]

  # Autocomplete resolvers.
  _unused_gravity_match_partners(match_type: String, page: Int = 1, size: Int = 5, term: String!): [DoNotUseThisPartner!] @deprecated(reason: \\"Use matchPartners\\")

  # Find partners by IDs
  _unused_gravity_partners(ids: [ID!]!): [DoNotUseThisPartner!]
}

# Autogenerated input type of RecordArtworkView
input RecordArtworkViewInput {
  artwork_id: String!

  # A unique identifier for the client performing the mutation.
  clientMutationId: String
}

# Autogenerated return type of RecordArtworkView
type RecordArtworkViewPayload {
  # Id of viewed artwork
  artworkId: String!

  # Id of viewed artwork
  artwork_id: String! @deprecated(reason: \\"Use artworkId\\")

  # A unique identifier for the client performing the mutation.
  clientMutationId: String
}
"
`;
